File: agent/cmc.py
"""CoinMarketCap Pro client — the deterministic data layer (Basic plan, in-credit).

Two load-bearing signals, both numeric (no free-text parsing in the control path):
  - per-token `trending` = 24h volume change %  (volume spiking = the name is heating up)
  - market `fear_greed`  = the CMC Fear & Greed index, logged as regime context

Key is read from env CMC_API_KEY or a 0600 key file — never hard-coded, never logged.
The qualitative Skill-Hub evidence packs (altcoin_kol_sentiment etc.) are a separate
submission/analyst artefact, deliberately kept OUT of the deterministic loop.
"""
from __future__ import annotations

import os
from pathlib import Path

import requests

from .config import CMC_IDS

BASE = "https://pro-api.coinmarketcap.com"


def _load_key() -> str:
    key = os.environ.get("CMC_API_KEY")
    if key:
        return key.strip()
    for p in (os.environ.get("CMC_KEY_FILE"), "~/.config/bnbagent/cmc_key"):
        if p and Path(p).expanduser().exists():
            return Path(p).expanduser().read_text().strip()
    raise RuntimeError("no CMC key (set CMC_API_KEY or ~/.config/bnbagent/cmc_key)")


class Cmc:
    def __init__(self, key: str | None = None, timeout: int = 30):
        self.key = key or _load_key()
        self.timeout = timeout

    def _get(self, path: str, params: dict) -> dict:
        r = requests.get(BASE + path, headers={"X-CMC_PRO_API_KEY": self.key,
                         "Accept": "application/json"}, params=params, timeout=self.timeout)
        j = r.json()
        st = j.get("status", {})
        if int(st.get("error_code") or 0) != 0:  # CMC returns "0" (str) or 0 (int) when OK
            raise RuntimeError(f"CMC {path}: {st.get('error_message')}")
        return j.get("data", {})

    def fear_greed(self) -> dict:
        """{'value': 0-100, 'classification': str} — market regime context."""
        d = self._get("/v3/fear-and-greed/latest", {})
        return {"value": int(d.get("value", 50)), "classification": d.get("value_classification", "")}

    def heat(self, ids: dict[str, int] | None = None) -> dict[str, dict[str, float]]:
        """{token: {'trending': volume_change_24h}} for the pinned meme ids."""
        ids = ids or CMC_IDS
        data = self._get("/v2/cryptocurrency/quotes/latest",
                         {"id": ",".join(str(i) for i in ids.values()), "convert": "USD"})
        by_id = {str(i): tok for tok, i in ids.items()}
        out: dict[str, dict[str, float]] = {}
        for cid, entry in data.items():
            tok = by_id.get(str(cid))
            if not tok:
                continue
            q = (entry.get("quote") or {}).get("USD", {})
            v = q.get("volume_change_24h")
            if v is not None:
                out[tok] = {"trending": float(v)}
        return out

File: rust/src/cmc.rs
//! CoinMarketCap Pro client — the deterministic data layer. One numeric signal feeds the loop:
//! per-token `trending` = 24h volume change %. The market Fear & Greed index is read as logged
//! context only. The key comes from env CMC_API_KEY or a 0600 key file — never hard-coded.

use std::collections::BTreeMap;
use std::path::PathBuf;

use crate::config::CMC_IDS;

const BASE: &str = "https://pro-api.coinmarketcap.com";

pub struct Cmc {
    key: String,
}

fn load_key() -> Result<String, String> {
    if let Ok(k) = std::env::var("CMC_API_KEY") {
        if !k.trim().is_empty() {
            return Ok(k.trim().to_string());
        }
    }
    let mut candidates: Vec<PathBuf> = vec![];
    if let Ok(p) = std::env::var("CMC_KEY_FILE") {
        candidates.push(PathBuf::from(p));
    }
    if let Ok(home) = std::env::var("HOME") {
        candidates.push(PathBuf::from(home).join(".config/bnbagent/cmc_key"));
    }
    for p in candidates {
        if let Ok(s) = std::fs::read_to_string(&p) {
            return Ok(s.trim().to_string());
        }
    }
    Err("no CMC key (set CMC_API_KEY or ~/.config/bnbagent/cmc_key)".into())
}

impl Cmc {
    pub fn new() -> Result<Cmc, String> {
        Ok(Cmc { key: load_key()? })
    }

    fn get(&self, path: &str, query: &[(&str, &str)]) -> Result<serde_json::Value, String> {
        let mut req = ureq::get(&format!("{BASE}{path}"))
            .set("X-CMC_PRO_API_KEY", &self.key)
            .set("Accept", "application/json");
        for (k, v) in query {
            req = req.query(k, v);
        }
        let j: serde_json::Value = req
            .call()
            .map_err(|e| format!("CMC {path}: {e}"))?
            .into_json()
            .map_err(|e| format!("CMC {path} body: {e}"))?;
        let code = j["status"]["error_code"].as_i64()
            .or_else(|| j["status"]["error_code"].as_str().and_then(|s| s.parse().ok()))
            .unwrap_or(0);
        if code != 0 {
            return Err(format!("CMC {path}: {}", j["status"]["error_message"]));
        }
        Ok(j["data"].clone())
    }

    /// {value, classification} — market regime context, logged not traded.
    pub fn fear_greed(&self) -> Result<(i64, String), String> {
        let d = self.get("/v3/fear-and-greed/latest", &[])?;
        Ok((
            d["value"].as_i64().unwrap_or(50),
            d["value_classification"].as_str().unwrap_or("").to_string(),
        ))
    }

    /// {token: {"trending": volume_change_24h}} for the pinned meme ids.
    pub fn heat(&self) -> Result<BTreeMap<String, BTreeMap<String, f64>>, String> {
        let ids: Vec<String> = CMC_IDS.iter().map(|(_, i)| i.to_string()).collect();
        let by_id: BTreeMap<String, &str> = CMC_IDS.iter().map(|(t, i)| (i.to_string(), *t)).collect();
        let data = self.get("/v2/cryptocurrency/quotes/latest", &[("id", &ids.join(",")), ("convert", "USD")])?;
        let mut out = BTreeMap::new();
        if let Some(obj) = data.as_object() {
            for (cid, entry) in obj {
                if let Some(tok) = by_id.get(cid) {
                    if let Some(v) = entry["quote"]["USD"]["volume_change_24h"].as_f64() {
                        out.insert(tok.to_string(), BTreeMap::from([("trending".to_string(), v)]));
                    }
                }
            }
        }
        Ok(out)
    }
}
